sympy/solvers/diophantine.py

```diff
@@ -177,11 +180,70 @@ def diophantine(eq, param=symbols("t", integer=True), syms=None):
         raise TypeError(filldedent('''
     Equation should be a polynomial with Rational coefficients.'''))
 
+    do_permute_signs = False
     try:
         # if we know that factoring should not be attempted, skip
         # the factoring step
         v, c, t = classify_diop(eq)
-        if t  == 'general_sum_of_squares':
+        if permute_sign:
+            len_var = len(v)
+            permute_signs_for = [
+                'general_sum_of_squares',
+                'general_sum_of_even_powers',
+                'general_pythagorean']
+            permute_signs_check = [
+                'homogeneous_ternary_quadratic',
+                'homogeneous_ternary_quadratic_normal',
+                'binary_quadratic']
+            if t in permute_signs_for:
+                do_permute_signs = True
+            elif t in (permute_signs_check):
+                # if all the variables in eq have even powers
+                # then do_permute_sign = True
+                if len_var == 3:
+                    var_mul = list(subsets(v, 2))
+                    # here var_mul is like [(x, y), (x, z), (y, z)]
+                    check_coeff = True
+                    a, b = symbols('a, b')
+                    var1_mul_var2 = map(lambda a, b: a * b, var_mul)
+                    for v1_mul_v2 in var1_mul_var2:
+                        try:
+                            coefficint = c[v1_mul_v2]
+                        except KeyError:
+                            coefficint = 0
+                        # if coeff(y*z), coeff(y*x), coeff(x*z) is not 0 then
+                        # check_coeff will be True and do_permute_sign will
+                        #  remain False.
+                        check_coeff = bool(check_coeff & coefficint)
+                    if not check_coeff:
+                        # means only x**2, y**2, z**2, const is present
+                        do_permute_signs = True
+                elif len_var == 2:
+                    var_mul = list(subsets(v, 2))
+                    # here var_mul is like [(x, y)]
+                    check_coeff = True
+                    x, y = symbols('x, y')
+                    global map
```

Comment:
Since `map` keyword has already been assigned here 
<https://github.com/Shekharrajak/sympy/blob/4e80f22ee715433a4e9512cc2b02e266283082ea/sympy/solvers/diophantine.py#L160> 
. Perhaps that name "map" should better be changed. It is not wise to 
use `global` keyword, I guess.

On 07/01/2016 05:11 PM, Shekhar Prasad Rajak wrote:
>
> In sympy/solvers/diophantine.py 
> <https://github.com/sympy/sympy/pull/11334#discussion_r69287067>:
>
> > +                            coefficint = c[v1_mul_v2]
> > +                        except KeyError:
> > +                            coefficint = 0
> > +                        # if coeff(y*z), coeff(y*x), coeff(x*z) is not 0 then
> > +                        # check_coeff will be True and do_permute_sign will
> > +                        #  remain False.
> > +                        check_coeff = bool(check_coeff & coefficint)
> > +                    if not check_coeff:
> > +                        # means only x**2, y**2, z**2, const is present
> > +                        do_permute_signs = True
> > +                elif len_var == 2:
> > +                    var_mul = list(subsets(v, 2))
> > +                    # here var_mul is like [(x, y)]
> > +                    check_coeff = True
> > +                    x, y = symbols('x, y')
> > +                    global map
>
> Without this line (|global map|). I got |unboundlocalerror| : map 
> local variable referenced before assignment. If someone knows find the 
> error then let me know.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/sympy/sympy/pull/11334/files/4e80f22ee715433a4e9512cc2b02e266283082ea#r69287067>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AHWt8YVKg978jNI0eUAco2mUibkRAs7Nks5qRPzngaJpZM4JDDvX>.
>